cpu/gd32v/periph/adc.c

```diff
@@ -0,0 +1,207 @@
+
+    /* enable the ADC module */
+    dev(line)->CTL1 |= ADC0_CTL1_ADCON_Msk;
+
+    /* configure the resolution */
+    dev(line)->OVSAMPCTL &= ~ADC0_OVSAMPCTL_DRES_Msk;
+    switch (res) {
+        case ADC_RES_12BIT:
+            dev(line)->OVSAMPCTL |= 0 << ADC0_OVSAMPCTL_DRES_Pos;
+            break;
+        case ADC_RES_10BIT:
+            dev(line)->OVSAMPCTL |= 1 << ADC0_OVSAMPCTL_DRES_Pos;
+            break;
+        case ADC_RES_8BIT:
+            dev(line)->OVSAMPCTL |= 2 << ADC0_OVSAMPCTL_DRES_Pos;
+            break;
+        case ADC_RES_6BIT:
+            dev(line)->OVSAMPCTL |= 3 << ADC0_OVSAMPCTL_DRES_Pos;
+            break;
+        default:
+            break;
+    }
+
+    /* check if this channel is an internal ADC channel, if so
+     * enable the internal temperature and Vref */
+    if (adc_config[line].chan == 16 || adc_config[line].chan == 17) {
+        dev(line)->CTL1 |= ADC0_CTL1_TSVREN_Msk;
+    }
+}
+
+static inline void done(adc_t line)
+{
+    /* disable the internal temperature and Vref */
+    dev(line)->CTL1 &= ~ADC0_CTL1_TSVREN_Msk;
+
+    /* disable the ADC module */
+    dev(line)->CTL1 &= ~ADC0_CTL1_ADCON_Msk;
+
+    periph_clk_dis(APB2, (RCU_APB2EN_ADC0EN_Msk << adc_config[line].dev));
+    mutex_unlock(&locks[adc_config[line].dev]);
+}
+
+int adc_init(adc_t line)
+{
+    uint32_t clk_div = 2;
+
+    /* check if the line is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+
+    /* lock and power-on the device */
+    prep(line, ADC_RES_12BIT);
+
+    /* configure the pin */
+    if (adc_config[line].pin != GPIO_UNDEF) {
+        gpio_init_analog(adc_config[line].pin);
+    }
+    /* set clock prescaler to get the maximal possible ADC clock value */
+    for (clk_div = 2; clk_div < 8; clk_div += 2) {
+        if ((CLOCK_CORECLOCK / clk_div) <= MAX_ADC_SPEED) {
+            break;
+        }
+    }
+    RCU->CFG0 &= ~(RCU_CFG0_ADCPSC_2_Msk);
+    RCU->CFG0 |= ((clk_div / 2) - 1) << RCU_CFG0_ADCPSC_2_Pos;
+
+    /* resets the selected ADC calibration registers */
+    dev(line)->CTL1 |= ADC0_CTL1_RSTCLB_Msk;
+    /* check the status of RSTCAL bit */
+    while (dev(line)->CTL1 & ADC0_CTL1_RSTCLB_Msk) {}
+
+    /* enable the selected ADC calibration process */
+    dev(line)->CTL1 |= ADC0_CTL1_CLB_Msk;
+    /* wait for the calibration to have finished */
+    while (dev(line)->CTL1 & ADC0_CTL1_CLB_Msk) {}
+
+    /* set all channels to maximum (239.5) cycles for best accuracy */
+    dev(line)->SAMPT0 |= 0x00ffffff;
+    dev(line)->SAMPT1 |= 0x3fffffff;
+    /* we want to sample one channel */
+    dev(line)->RSQ0 = 1 << ADC0_RSQ0_RL_Pos;
+    /* start sampling from software */
+    dev(line)->CTL1 |= ADC0_CTL1_ETERC_Msk | ADC0_CTL1_ETSRC_Msk;
+
+    /* check if the internal channels are configured to use ADC0 */
+    if (adc_config[line].chan == 16 || adc_config[line].chan == 17) {
+        if (dev(line) != ADC0) {
+            return -3;
+        }
```

Comment:
```suggestion
        assert(dev(line) == ADC0);
```
since that'd be an error in the board config, let this fail loudly.

Reply:
That's just the code from STM32F1. Using `assert` here makes perfect sense, especially since the API does not define -3 as a possible return value. In fact, I had already thought about an `assert`.